src/libmongoc/src/mongoc/mongoc-uri.c

```diff
@@ -1424,6 +1426,25 @@ mongoc_uri_parse (mongoc_uri_t *uri, const char *str, bson_error_t *error)
    }
 
    before_slash = scan_to_unichar (str, '/', "", &tmp);
+   if (!before_slash) {
+      // Handle cases of optional delimiting slash
+
+      // Skip any "?" that exist is the userpass
+      userpass = scan_to_unichar (str, '@', "", &tmp);
+      if (!userpass) {
+         // If none found, we can safely check for "?" indicating beginning of options
+         before_slash = scan_to_unichar (str, '?', "", &tmp);
+      } else {
+         // Otherwise, see if options exist after username/password and concatenate result
+         hostname = scan_to_unichar (tmp, '?', "", &tmp);
+
+         if (hostname) {
+            before_slash = bson_strdup (str);
+            before_slash[strlen (userpass) + strlen (hostname)] = '\0';
```

Comment:
Open for suggestions on how to accomplish this in a more efficient way. I tried messing around with `bson_strncpy` but couldn't get it to work.

Reply:
`scan_to_unichar` returns a pointer past the last scanned character via `&tmp` when the return value is not null. Therefore, the length of the scanned string given `scan_to_unichar (begin, ..., &end)` can be computed as `end - begin`.

```c
// Handle cases of optional delimiting slash

// Skip any "?"s that exist in the userpass
userpass = scan_to_unichar (str, '@', "", &tmp);
if (!userpass) {
   // If none found, safely check for "?" indicating beginning of options
   before_slash = scan_to_unichar (str, '?', "", &tmp);
} else {
   const size_t userpass_len = (size_t) (tmp - str);

   // Otherwise, see if options exist after userpass and concatenate result
   hostname = scan_to_unichar (tmp, '?', "", &tmp);

   if (hostname) {
      const size_t hostname_len = (size_t) (tmp - str) -userpass_len;

      before_slash = bson_strndup (str, userpass_len + hostname_len);
   }
}
```